



<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
      
        <link rel="canonical" href="https://lamdv.github.io/Scientific Journey/Report/report/">
      
      
        <meta name="author" content="D">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.0.4, mkdocs-material-4.4.3">
    
    
      
        <title>Report on combining (Stack) C+T predictor</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/application.30686662.css">
      
      
    
    
      <script src="../../../assets/javascripts/modernizr.74668098.js"></script>
    
    
      
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../../../assets/fonts/material-icons.css">
    
    
    
      
    
    
  </head>
  
    <body dir="ltr">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="__github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#methods" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="https://lamdv.github.io/" title="Liam" class="md-header-nav__button md-logo">
          
            <i class="md-icon"></i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            <span class="md-header-nav__topic">
              Liam
            </span>
            <span class="md-header-nav__topic">
              
                Report on combining (Stack) C+T predictor
              
            </span>
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
          
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  

<a href="https://github.com/lamdv/lamdv.github.io/" title="Go to repository" class="md-source" data-md-source="github">
  
    <div class="md-source__icon">
      <svg viewBox="0 0 24 24" width="24" height="24">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    GitHub
  </div>
</a>
          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main" role="main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="https://lamdv.github.io/" title="Liam" class="md-nav__button md-logo">
      
        <i class="md-icon"></i>
      
    </a>
    Liam
  </label>
  
    <div class="md-nav__source">
      


  

<a href="https://github.com/lamdv/lamdv.github.io/" title="Go to repository" class="md-source" data-md-source="github">
  
    <div class="md-source__icon">
      <svg viewBox="0 0 24 24" width="24" height="24">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href="../../.." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../../../CV/" title="Curriculum Vitae" class="md-nav__link">
      Curriculum Vitae
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3" checked>
    
    <label class="md-nav__link" for="nav-3">
      Scientific Journey
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-3">
        Scientific Journey
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-3-1" type="checkbox" id="nav-3-1">
    
    <label class="md-nav__link" for="nav-3-1">
      Class Note @ENSIMAG
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="2">
      <label class="md-nav__title" for="nav-3-1">
        Class Note @ENSIMAG
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../../../Master/class notes/Fundamental ML/notes/" title="Fundamental ML" class="md-nav__link">
      Fundamental ML
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../intro/" title="Hành trình khoa học: Python và R" class="md-nav__link">
      Hành trình khoa học: Python và R
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../day-one-ish/" title="(Những) ngày đầu tiên" class="md-nav__link">
      (Những) ngày đầu tiên
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../day-two-documentation-markdown/" title="Ghi chú bằng Markdown" class="md-nav__link">
      Ghi chú bằng Markdown
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../xgboost_c+t/" title="XGBoost's objective function comparison" class="md-nav__link">
      XGBoost's objective function comparison
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        Viable method to combine C+T predictors for enhanced prediction power
      </label>
    
    <a href="./" title="Viable method to combine C+T predictors for enhanced prediction power" class="md-nav__link md-nav__link--active">
      Viable method to combine C+T predictors for enhanced prediction power
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#methods" class="md-nav__link">
    Methods
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#required-packages" class="md-nav__link">
    Required packages:
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dataset" class="md-nav__link">
    Dataset
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ct-step" class="md-nav__link">
    C+T step
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-treatment-oversampling" class="md-nav__link">
    Data treatment: Oversampling
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#method-1-sct" class="md-nav__link">
    Method 1: SCT
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#method-2-xgboost" class="md-nav__link">
    Method 2: XGBoost
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#method-3-lassosum" class="md-nav__link">
    Method 3: Lassosum
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#method-4-random-forest" class="md-nav__link">
    Method 4: Random Forest
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#method-5-nerual-net-keras" class="md-nav__link">
    Method 5: Nerual net (Keras)
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

        
      </ul>
    </nav>
  </li>

    
      
      
      


  <li class="md-nav__item md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-4" type="checkbox" id="nav-4">
    
    <label class="md-nav__link" for="nav-4">
      Thought
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-4">
        Thought
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../../../Thought/the-death-of-expertise/" title="The Death of Expertise" class="md-nav__link">
      The Death of Expertise
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../../../Thought/on-writing/" title="Về chuyện viết lách" class="md-nav__link">
      Về chuyện viết lách
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#methods" class="md-nav__link">
    Methods
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#required-packages" class="md-nav__link">
    Required packages:
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dataset" class="md-nav__link">
    Dataset
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ct-step" class="md-nav__link">
    C+T step
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#data-treatment-oversampling" class="md-nav__link">
    Data treatment: Oversampling
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#method-1-sct" class="md-nav__link">
    Method 1: SCT
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#method-2-xgboost" class="md-nav__link">
    Method 2: XGBoost
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#method-3-lassosum" class="md-nav__link">
    Method 3: Lassosum
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#method-4-random-forest" class="md-nav__link">
    Method 4: Random Forest
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#method-5-nerual-net-keras" class="md-nav__link">
    Method 5: Nerual net (Keras)
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/lamdv/lamdv.github.io/edit/source/docs/Scientific Journey/Report/report.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                  <h1>Viable method to combine C+T predictors for enhanced prediction power</h1>
                
                <p>In this document I investigate multiple method to calculate PRS. Some of them are based on combining multiple C+T predictors to enhence the performance of C+T. They are based on the work of <a href="https://privefl.github.io/bigsnpr/articles/SCT.html">Florian Privé</a>. <sup id="fnref:1"><a class="footnote-ref" href="#fn:1">1</a></sup></p>
<h2 id="methods">Methods</h2>
<p>In this RMD documents there are 5 methods demonstrated:</p>
<ul>
<li>SCT</li>
<li>XGBoost (with SCT <script type="math/tex">1^{st}</script> layer)</li>
<li>Lassosum</li>
<li>Random Forest (with SCT <script type="math/tex">1^{st}</script> layer)</li>
<li>Neural Net (with SCT <script type="math/tex">1^{st}</script> layer)</li>
</ul>
<p>More promising (SCT, XGBoost, Lassosum, Random Forest) are place first, and Neural Net are included for reference.</p>
<p>For some reason I can not get Keras to work with SCT first layer (always return AUC of 0.5). This can either be a problem of the target function (I have yet to find a good function that operate well), or a bug within Keras (unlikely). Or perhap the Euclid distance after C+T layer are too small for neuralnet - that's why regression based method work better?</p>
<h2 id="required-packages">Required packages:</h2>
<ul>
<li>bigstatsr</li>
<li>bigsnpr</li>
<li>lassosum</li>
<li>data.table</li>
<li>xgboost</li>
<li>ranger</li>
<li>keras</li>
</ul>
<pre><code class="r">library(bigstatsr)
library(bigsnpr)
library(lassosum)
library(data.table)
library(xgboost)
library(ranger)
library(keras)
</code></pre>

<h2 id="dataset">Dataset</h2>
<p><em>Please change working direction in <code>directory</code></em></p>
<p>I use the "Simus" simulated dataset provided by Florian. This dataset contain ~650,000 SNPs in 2 chromosomes. 20% of them are cases and the rest 80% are controls.</p>
<pre><code class="r">setwd(directory) 
# Train data
sumstats &lt;- bigreadr::fread2(file_sumstat)
# snp_readBed(paste(file_train, &quot;bed&quot;, sep='.'))
train &lt;- snp_attach(paste(file_train, &quot;rds&quot;, sep='.'))
G.train &lt;- train$genotypes
CHR &lt;- train$map$chromosome
POS &lt;- train$map$physical.pos
NCORES &lt;- nb_cores()
lpval &lt;- -log10(sumstats$pval)
y.train &lt;- train$fam$affection

# Test data
# snp_readBed(paste(file_test, &quot;bed&quot;, sep='.'))
test &lt;- snp_attach(paste(file_test, &quot;rds&quot;, sep='.'))
G.test &lt;- test$genotypes
# 
</code></pre>

<h2 id="ct-step">C+T step</h2>
<p>Generating a matrix of C+T with varied clumping radius and <script type="math/tex">p-value</script> thresholds.</p>
<p>
<script type="math/tex">\beta s</script> and <script type="math/tex">p-values</script> are aquired in sumstats file (see <code>sumstats.txt</code>)</p>
<pre><code class="r">all_keep &lt;- snp_grid_clumping(G.train, CHR, POS, lpval, ncores = NCORES)
PRS &lt;-snp_grid_PRS(G.train,all_keep = all_keep, betas = sumstats$beta,lpval)
PRS_test &lt;- snp_grid_PRS(G.test, all_keep = all_keep, betas = sumstats$beta,lpval)
</code></pre>

<p>Other than SCT, other methods require DataFrame or R Matrix than FBM to manage data. I converted the FBM output of C+T to DataFrame and change the corresponding columns.</p>
<h2 id="data-treatment-oversampling">Data treatment: Oversampling</h2>
<p>To oversampling I extract all positive (<em>affection = 1</em>), and append them a number of times. For this simulated dataset, I enhanced the number of cases 10 times, thus the ratio become ~70:30 (from 20:80).</p>
<pre><code class="r">y_case &lt;- which(train$fam$affection %in% c(1))
for (i in seq(1, 10)){
  y.train &lt;- append(y.train, train$fam$affection[y_case])
  ds &lt;- rbind(ds, PRS[y_case,])
}
ds.FBM &lt;- as_FBM(ds)
</code></pre>

<h2 id="method-1-sct">Method 1: SCT</h2>
<p>No parameters required. Combining (Sparse Logistic Regression) C+T predictors for best result.</p>
<pre><code class="r"># Reference
# M &lt;- snp_grid_stacking(multi_PRS = ds.FBM, y.train = y.train, ncores = NCORES)
M &lt;- snp_grid_stacking(multi_PRS = PRS, y.train = train$fam$affection)
beta &lt;- as_FBM(matrix(M$beta.G))
pred.SCT &lt;- big_prodMat(G.test, beta)
#AUC(pred = pred.SCT, test$fam$affection)
</code></pre>

<pre><code class="r">AUC(pred = pred.SCT, test$fam$affection)
</code></pre>

<pre><code>## [1] 0.7816448
</code></pre>

<h2 id="method-2-xgboost">Method 2: XGBoost</h2>
<p>XGBoost provide slightly better result compare to SCT. However this result heavily depend on booster and objective selection. Experiments shown that <strong>gblinear</strong> booster with <strong>count:poisson</strong> objective give the best results.</p>
<p>For experimentation, I run 9 different <em>max_depth</em> and 9 <em>nrounds</em>. Each combination is repeated 10 times and the mean value calculated to rule out the randomess of <em>gblinear</em> booster</p>
<pre><code class="r">AUCs &lt;- matrix(,nrow = 10, ncol = 10)
for (i in seq(2 : 10)){
  for (j in seq(2: 10)) {
    temp &lt;- c()
    for(t in seq(1:10)){
      bstSparse &lt;- xgboost(data = ds[], 
                           label = y.train, 
                           booster=&quot;gblinear&quot;,
                           max_depth = i, 
                           eta = 1, 
                           nthread = 2, 
                           nrounds = j, 
                           lambda = 0.1,
                           objective = &quot;count:poisson&quot;
                           )
      pred &lt;- predict(bstSparse, PRS_test[])
      temp &lt;- append(temp,AUC(pred = pred, test$fam$affection))
    }
    AUCs[i,j] &lt;- mean(temp)
  }
}
</code></pre>

<pre><code class="r">AUCs
</code></pre>

<pre><code>##            [,1]      [,2]      [,3]      [,4]      [,5]      [,6]
##  [1,] 0.7810081 0.7814691 0.7819848 0.7820644 0.7821634 0.7819551
##  [2,] 0.7815161 0.7816789 0.7819097 0.7820451 0.7820130 0.7820510
##  [3,] 0.7811385 0.7817903 0.7820209 0.7821792 0.7821172 0.7820448
##  [4,] 0.7812592 0.7814580 0.7820362 0.7819456 0.7820135 0.7819257
##  [5,] 0.7811024 0.7814672 0.7818094 0.7819867 0.7820772 0.7820431
##  [6,] 0.7813253 0.7816065 0.7818425 0.7822446 0.7821884 0.7819205
##  [7,] 0.7810239 0.7817146 0.7817587 0.7819281 0.7822180 0.7819721
##  [8,] 0.7811480 0.7820816 0.7818343 0.7820383 0.7820944 0.7819818
##  [9,] 0.7815687 0.7814758 0.7816728 0.7818929 0.7820611 0.7820187
## [10,]        NA        NA        NA        NA        NA        NA
##            [,7]      [,8]      [,9] [,10]
##  [1,] 0.7817276 0.7816945 0.7814971    NA
##  [2,] 0.7819033 0.7817853 0.7816005    NA
##  [3,] 0.7818634 0.7816886 0.7814594    NA
##  [4,] 0.7819806 0.7818078 0.7813999    NA
##  [5,] 0.7818562 0.7816829 0.7814363    NA
##  [6,] 0.7819113 0.7817263 0.7814301    NA
##  [7,] 0.7818913 0.7817029 0.7815198    NA
##  [8,] 0.7819087 0.7816597 0.7814085    NA
##  [9,] 0.7817407 0.7817544 0.7815585    NA
## [10,]        NA        NA        NA    NA
</code></pre>

<pre><code class="r">boxplot(AUCs)
</code></pre>

<p><img alt="" src="../report_fig/result_xgboost-1.png" /><!-- --></p>
<h2 id="method-3-lassosum">Method 3: Lassosum</h2>
<p>Lassosum use L1 regularization to better fit the regression model.</p>
<p>For basic L1 regularization of Linear Regression, 2 metaparameters are required: a learning rate <script type="math/tex">\alpha</script> and a regularization parameter <script type="math/tex">\lambda</script>. However, <em>lassosum</em> does not require these parameters, as the software automatically scan the parameters space and select the best <script type="math/tex">\alpha</script> and <script type="math/tex">\lambda</script>
</p>
<pre><code class="r">setwd(directory) 
cor &lt;- p2cor(p = sumstats$pval, n = 8000, sign=sumstats$beta)

out &lt;- lassosum.pipeline(cor =cor, chr=sumstats$chromosome, pos = sumstats$physical.pos, 
                         A1 = sumstats$allele1, A2 = sumstats$allele2,
                         ref.bfile = file_train, test.bfile = file_test,
                         LDblocks = LDblocks)

v &lt;- validate(out)

out2 &lt;- subset(out, s=v$best.s, lambda = v$lambda)
v2 &lt;- validate(out2)
v2$best.validation.result
</code></pre>

<pre><code class="r">AUC(v$best.pgs, v$pheno)
</code></pre>

<pre><code>## [1] 0.7308175
</code></pre>

<h2 id="method-4-random-forest">Method 4: Random Forest</h2>
<p>Random forest also give good AUC (above 70%, but less than other methods except for <em>keras</em>). The result highly dependen on the number of trees (more trees equals more convergence), and the learning rate also contribute to an lesser extend.</p>
<p>The code in this session will sweep through the parameters: 1 to 10 trees and alpha from 0.1 to 1 (0.1 increment).</p>
<pre><code class="r">PRS_test.df &lt;- as.data.frame(PRS_test[])

f &lt;- as.formula(paste(&quot;y ~&quot;, paste(cols[!cols %in% &quot;PRS.df&quot;], collapse = &quot; + &quot;)))

for (j in seq(1 : 10)){
  for (i in seq(1: 10)) {
    rf &lt;- ranger(f, data = PRS.df, num.trees = j*100, verbose = TRUE, write.forest = TRUE, alpha = (i/10))
    pred &lt;- predict(rf, PRS_test.df)
    AUCs[i,j] &lt;- AUC(pred = pred$predictions, test$fam$affection)
  }
}
</code></pre>

<pre><code class="r">AUCs
</code></pre>

<pre><code>##            [,1]      [,2]      [,3]      [,4]      [,5]      [,6]
##  [1,] 0.7532205 0.7535048 0.7564268 0.7501722 0.7507132 0.7515333
##  [2,] 0.7440113 0.7530198 0.7540833 0.7532506 0.7518778 0.7546351
##  [3,] 0.7545732 0.7500895 0.7514272 0.7556325 0.7545590 0.7539930
##  [4,] 0.7549227 0.7515300 0.7500226 0.7548174 0.7503537 0.7543692
##  [5,] 0.7488554 0.7515392 0.7508904 0.7535265 0.7544863 0.7536828
##  [6,] 0.7501346 0.7512984 0.7525976 0.7547204 0.7520592 0.7536302
##  [7,] 0.7496137 0.7509414 0.7527130 0.7549386 0.7512415 0.7553466
##  [8,] 0.7435724 0.7513862 0.7545431 0.7513971 0.7525416 0.7559703
##  [9,] 0.7513695 0.7512825 0.7537756 0.7543274 0.7507516 0.7562646
## [10,] 0.7473798 0.7476314 0.7535859 0.7539036 0.7543751 0.7537062
##            [,7]      [,8]      [,9]     [,10]
##  [1,] 0.7541560 0.7521771 0.7544930 0.7532481
##  [2,] 0.7547923 0.7526261 0.7558700 0.7539353
##  [3,] 0.7526620 0.7529212 0.7531896 0.7540800
##  [4,] 0.7532573 0.7556534 0.7528978 0.7538868
##  [5,] 0.7518979 0.7547931 0.7558106 0.7535190
##  [6,] 0.7544119 0.7547555 0.7558792 0.7548132
##  [7,] 0.7524614 0.7543283 0.7537857 0.7549896
##  [8,] 0.7533702 0.7535374 0.7533777 0.7520208
##  [9,] 0.7569953 0.7558482 0.7558123 0.7530600
## [10,] 0.7538342 0.7547547 0.7521612 0.7539888
</code></pre>

<pre><code class="r">boxplot(AUCs)
</code></pre>

<p><img alt="" src="../report_fig/result_rf-1.png" /><!-- --></p>
<h2 id="method-5-nerual-net-keras">Method 5: Nerual net (Keras)</h2>
<p>For some reason, <em>keras</em> cannot return good result; the prediction always biased to controls. Even with adjusted weights and oversampling, the bias is still there and AUC is always <script type="math/tex">0.5</script>. I included it here in case I figure it out in the future.</p>
<p>In this example, I put a 10:1 weight ratio between case and control.</p>
<pre><code class="r">y.train &lt;- train$fam$affection

model &lt;- keras_model_sequential() 
model %&gt;%
 layer_dense(units = 1000, activation = &quot;softmax&quot;, input_shape = c(ncol(ds)), kernel_regularizer=regularizer_l1(0.02)) %&gt;%
 # layer_dense(units = 1000, activation = &quot;softmax&quot;, input_shape = c(2800)) %&gt;%
 # layer_dropout(rate=0.1) %&gt;%
 layer_dense(units = 2, activation = &quot;softmax&quot;, kernel_regularizer=regularizer_l1(0.01))

model %&gt;% compile(
 #loss = 'categorical_crossentropy',
 loss = 'mean_absolute_percentage_error',
 optimizer = 'SGD',
 metrics = c('accuracy')
)
y.train &lt;-  to_categorical(y.train)

history &lt;- model$fit(
 ds, y.train,
 class_weight = list(1, 10),
 epochs = as.integer(10), 
 batch_size = as.integer(28)
)
AUC(predict_classes(model, PRS_test[]), test$fam$affection)
</code></pre>

<pre><code>## [1] 0.5
</code></pre>

<div class="footnote">
<hr />
<ol>
<li id="fn:1">
<p>Original Rmd documment provided <a href="https://github.com/lamdv/Script/blob/master/Report/report.rmd">here</a>&#160;<a class="footnote-backref" href="#fnref:1" title="Jump back to footnote 1 in the text">&#8617;</a></p>
</li>
</ol>
</div>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../../xgboost_c+t/" title="XGBoost's objective function comparison" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                XGBoost's objective function comparison
              </span>
            </div>
          </a>
        
        
          <a href="../../../Thought/the-death-of-expertise/" title="The Death of Expertise" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                The Death of Expertise
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../../assets/javascripts/application.ac79c3b0.js"></script>
      
      <script>app.initialize({version:"1.0.4",url:{base:"../../.."}})</script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
    
  </body>
</html>