
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="canonical" href="https://lamdv.github.io/Scientific%20Journey/Sequence%20visualization%20with%20matplotlib/">
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.1.2, mkdocs-material-7.0.6">
    
    
      
        <title>Visualizing sequence data with Matplotlib - Liam</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.2c0c5eaf.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.7fa14f5b.min.css">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
    
      
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="none" data-md-color-accent="none">
      
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#visualizing-sequence-data-with-matplotlib" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Liam" class="md-header__button md-logo" aria-label="Liam" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Liam
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Visualizing sequence data with Matplotlib
            
          </span>
        </div>
      </div>
    </div>
    <div class="md-header__options">
      
    </div>
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" data-md-state="active" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/lamdv/lamdv.github.io/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Liam" class="md-nav__button md-logo" aria-label="Liam" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    Liam
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/lamdv/lamdv.github.io/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05L244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../CV/" class="md-nav__link">
        Curriculum Vitae
      </a>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_3" type="checkbox" id="__nav_3" >
      
      <label class="md-nav__link" for="__nav_3">
        Scientific Journey
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Scientific Journey" data-md-level="1">
        <label class="md-nav__title" for="__nav_3">
          <span class="md-nav__icon md-icon"></span>
          Scientific Journey
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../intro/" class="md-nav__link">
        Hành trình khoa học: Python và R
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../day-one-ish/" class="md-nav__link">
        (Những) ngày đầu tiên
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../day-two-documentation-markdown/" class="md-nav__link">
        Ghi chú bằng Markdown
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../xgboost_c%2Bt/" class="md-nav__link">
        XGBoost's objective function comparison
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../Report/report/" class="md-nav__link">
        Viable method to combine C+T predictors for enhanced prediction power
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../machine%20learning%20machine%20learning_en/" class="md-nav__link">
        Model Selection: Machine learning Machine Learning - Part 1
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
      
      
      

  
  
  
    
    <li class="md-nav__item md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_4" type="checkbox" id="__nav_4" >
      
      <label class="md-nav__link" for="__nav_4">
        Thought
        <span class="md-nav__icon md-icon"></span>
      </label>
      <nav class="md-nav" aria-label="Thought" data-md-level="1">
        <label class="md-nav__title" for="__nav_4">
          <span class="md-nav__icon md-icon"></span>
          Thought
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../Thought/the-death-of-expertise/" class="md-nav__link">
        The Death of Expertise
      </a>
    </li>
  

          
            
  
  
  
    <li class="md-nav__item">
      <a href="../../Thought/on-writing/" class="md-nav__link">
        Về chuyện viết lách
      </a>
    </li>
  

          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#the-dirty-way" class="md-nav__link">
    The dirty way
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#the-idea" class="md-nav__link">
    The idea
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#prequisite" class="md-nav__link">
    Prequisite
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#reading-stuff" class="md-nav__link">
    Reading stuff...
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#now-the-fun-part-actually-drawing-the-msa" class="md-nav__link">
    Now, the fun part: Actually drawing the MSA.
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/lamdv/lamdv.github.io/edit/source/docs/Scientific Journey/Sequence visualization with matplotlib.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <h1 id="visualizing-sequence-data-with-matplotlib">Visualizing sequence data with Matplotlib</h1>
<h5 id="the-dirty-way"><em>The dirty way</em></h5>
<p>There are probably many packages that take multiple sequences alignment and draw a pretty nice little visualization. However as a software engineer and bioinformatician, our time is probably better spend on algorithm than learning new packages, especially when it pretty one off. Furthermore, sometime it is important to have more control over our graph, including alignning elements, incorporate different metrics...</p>
<p>In this tutorial we will try to visualize a short aa sequence (75 residues) with conservation calculated by norm distance to uniform distribution (it's not the best way to calculate conservation but... it's quick and dirty). Finally we display a concensus sequence, which is the most relevant amino acid at each residue. Again this is not biologically correct, but that beside the point. You should be able quickly adapt this script to work on different data and different metrics</p>
<h2 id="the-idea">The idea</h2>
<p>The idea behind this notebook is as we plot statistic at each residue, we draw the sequence directly underneath it. The reason this method even work is matplotlib draw on the figure in "ticks space", as in the distance on the plot is calculated in data steps.</p>
<p>So if we have a bar graph, we can quickly align vertically a text object simply by putting <code>plot.text()</code> at exactly x. Conversely of y, we need to calculate an offset from the bottom of the graph to the position we want to draw our text/residue/nucleotide acid. That simple.</p>
<h2 id="prequisite">Prequisite</h2>
<ol>
<li>The necessary libraries
We need the following libraries: matplotlib (duh), BioPython and NumPy. BioPython is needed to effectively manage our MSA, although with parser and hard work you can read directly from fasta file. NumPy is needed to interface with matplotlib, and in general, who would not include NumPy in their tutorial these day</li>
</ol>
<pre><code class="language-python">import numpy as np, Bio, matplotlib.pyplot as plot
from matplotlib.ticker import FormatStrFormatter
from Bio import SeqIO

%matplotlib inline
</code></pre>
<ol>
<li>In order to display the sequence in a format we familiar with we need 2 things. First, because we intent to visualize our sequences in a sequence of boxes, each box represent a residue and we want the boxes to be align, we need a monospacing font family</li>
</ol>
<pre><code class="language-python">font = {'family': 'monospace',
        'color':  'darkred',
        'weight': 'normal',
        'size': 16,
        }
</code></pre>
<p>Secondly we need a color palette. There are 20 amino acid and one color for gap, which is displayed as a dash. The color I use here is a color palette used by RasMol package (holy f it's old)</p>
<pre><code class="language-python">palette = ['#C8C8C8', '#145AFF', '#00DCDC', '#E60A0A', '#E6E600', '#00DCDC', '#E60A0A', '#EBEBEB', '#8282D2', '#0F820F', '#0F820F', '#145AFF', '#E6E600', '#3232AA', '#DC9682', '#FA9600', '#FA9600', '#B45AB4', '#3232AA', '#0F820F', '#FFFFFF']
</code></pre>
<p>Finally we have a string of all possible amino acid + the gap</p>
<pre><code class="language-python">aa = 'ARNDCQEGHILKMFPSTWYV-'
</code></pre>
<h2 id="reading-stuff">Reading stuff...</h2>
<p>First thing first: we read our MSA with the most basic numbers: length and number of sequences</p>
<pre><code class="language-python">fasta_file = 'src/Inputs/example/1atzA.fas'
msa = list(SeqIO.parse(fasta_file,'fasta'))
L = len(msa[0].seq)
N = len(msa)
</code></pre>
<p>We calculate the frequency of each amino acid occurs at each residue next. We also calucate the concensus sequence at this stage</p>
<pre><code class="language-python">freq = np.zeros([L,21])
concensus = np.zeros(L)
for i in range(0,N):
    for j in range(0,L):
        j_aa = aa.find(msa[i].seq[j])
        freq[j,j_aa] = freq[j,j_aa] + 1
for i in range(0, L):
    concensus[i] = freq[i].argmax()
</code></pre>
<p>The algorithm is simple. </p>
<p>Next come the conservation. The formular for conservation is:
<script type="math/tex; mode=display">
C_i = \sum_{j}^{j=N}\sqrt{(\frac{f_j}{N} - 0.05)^2}
</script>
</p>
<p>For <script type="math/tex">C_i</script> is the conservation value of <script type="math/tex">i^{th}</script> residue.</p>
<p>Why 0.05? Because it's 1/20, and guess how many amino acid there is.</p>
<p><em>(of course there are 21 possible values but this function wrong enough lol)</em></p>
<pre><code class="language-python">conservation = np.sqrt(np.sum((np.square(freq/N - 0.05)),axis=1))
</code></pre>
<h2 id="now-the-fun-part-actually-drawing-the-msa">Now, the fun part: Actually drawing the MSA.</h2>
<p>Because there is no native way to draw all the thing that go into this visualization, it look a bit messy. And also because all of this is quick, dirty, this will probably not the best way to do thing. But it's a way.</p>
<p><em>Notes: I put </em><code>figure</code><em> in places where I want to draw the figure </em><em>for demonstration purpose</em><em>. You may want to remove them when you recombine the code boxes in your notebook, otherwise it will draw multiple graph. On itself.</em></p>
<p>First, we create  the figure with its axes. The figure is configured to be ultra wide, because the sequences are long <em>wink wink</em>.</p>
<pre><code class="language-python">figure = plot.figure(figsize=(20,2))
axes = plot.axes([0,0,1,1]);
plot.close()
</code></pre>
<p>Now, to draw the conservation graph. It's a simple bargraph which display at each residue what is the conservation rate according to the formular provided previously.</p>
<p>Reminder: to call bargraph: <code>axes.bar(range,data,**args)</code>. It is also important to have <code>align='edge', linewidth=0</code> for alignment purpose. <code>align='edge'</code> make the bar aligned on the right of the ticks, and <code>linewidth=0</code> make the spacing between ticks correct.</p>
<pre><code class="language-python">axes.bar(range(0,L),conservation, align='edge', linewidth = 0, color = 'red')
axes.set_ylabel('Conservation')
figure
</code></pre>
<p><img alt="svg" src="Sequence%20visualization%20with%20matplotlib_files/Sequence%20visualization%20with%20matplotlib_21_0.svg" /></p>
<p>Next step we draw the sequences. We choose randomly 5 sequences from the MSA, because to display all 1000 MSA of this dataset is simply ... unrealistic. We also calculate the spacing between each sequences, as well as spacing between the bottom of the conservation graph to the first sequence displayed.</p>
<p>The spacing scale is calculated by taking the limit of y and divided by 6. Why? Because it work for our particular font and spacing...</p>
<pre><code class="language-python">spacing_scale = axes.get_ylim()[1]/6
spacing = spacing_scale*2

seq_display = np.sort(np.random.randint(0,N,[5]))
</code></pre>
<p>So now, where to draw our sequence? For each sequence we calculate the y-offset from <script type="math/tex">(0, 0)</script> position in data space. Then the x position of each residue is simply its position in the sequence... Prety neat indeed.</p>
<p>The following code block go like: For each sequence, calculate the basic position (0,posit), where posit = the place of sequence times the scale minus an offset. <code>axes.text()</code> is called to draw the sequence title. Finally, each residue is draw by <code>axes.text()</code> with the x position is the position of the residue itself, and a bbox draw over the text. We feed the bbox the color according to the palette, and an <code>alpha=0.5</code> so it's see through.</p>
<p>Oh, and remember to add <code>fontdict=font</code> for our monospacing font set</p>
<pre><code class="language-python">for j in seq_display:
    posit = -float(np.where(seq_display == j)[0]) * spacing_scale - spacing
    axes.text(-5,posit, &quot;Seq &quot;+(str(j+1)))
    for i in range(0, L):
        # print(i/20,j/4)
        axes.text(float(i),posit, msa[j].seq[i],
            bbox=dict(facecolor=palette[aa.find(msa[j].seq[i])], 
            alpha=0.5),fontdict=font)
figure
</code></pre>
<p><img alt="svg" src="Sequence%20visualization%20with%20matplotlib_files/Sequence%20visualization%20with%20matplotlib_25_0.svg" /></p>
<p>Finally we add the concensus sequence:</p>
<pre><code class="language-python">posit = posit - spacing
axes.text(-5,posit, &quot;Concensus&quot;)
for i in range(0, L):
    # print(i/20,j/4)
    axes.text(float(i),posit, 'ARNDCQEGHILKMFPSTWYV-'[int(concensus[i])] ,
                bbox=dict(facecolor=palette[int(concensus[i])], 
                alpha=0.5),fontdict=font)
figure
</code></pre>
<p><img alt="svg" src="Sequence%20visualization%20with%20matplotlib_files/Sequence%20visualization%20with%20matplotlib_27_0.svg" /></p>
<p>Et voila! There you go, a quick and dirty way to visualize your sequence dataset. The code is (hopefully) simple enough so you can expanse on your own to other type of biological sequence data (Genomics, Trans...), and different graph can also be used as basis for visualizing your MSA. This, however, is not scalable to larger sequences. You will have to experiment on how best to show your data, but that's the fun of bioinfomatics! And now you have all the controls to do so.</p>
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Made with
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs
        </a>
        
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing"}, "search": "../../assets/javascripts/workers/search.fb4a9340.min.js", "version": null}</script>
    
    
      <script src="../../assets/javascripts/bundle.a1c7c35e.min.js"></script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
    
  </body>
</html>