---
title: "Report on combining (Stack) C+T predictor"
author: "DANG Vu-Lam"
output:
  html_document:
    keep_md: true
editor_options: 
  chunk_output_type: inline
---

```{r, setup, include=FALSE} 
knitr::opts_chunk$set(fig.width=12, fig.height=8, fig.path='Figs/', warning = FALSE)
knitr::opts_chunk$set(fig.path = 'report_fig/')
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(cache.lazy = FALSE)

directory <- "../simus"
file_sumstat <- "sumstats.txt"
file_train <- "data_train"
file_test <- "data_test"
LDblocks <- "EUR.hg19"
```

## Methods

In this RMD documents there are 5 methods demonstrated:

- SCT
- XGBoost (with SCT 1\hat(st) layer)
- Lassosum
- Random Forest (with SCT 1\hat(st) layer)
- Neural Net (with SCT 1\hat(st) layer)

More promising (SCT, XGBoost, Lassosum, Random Forest) are place first, and Neural Net are included for reference.

For some reason I can not get Keras to work with SCT first layer (always return AUC of 0.5). This can either be a problem of the target function (I have yet to find a good function that operate well), or a bug within Keras (unlikely). Or perhap the Euclid distance after C+T layer are too small for neuralnet - that's why regression based method work better?

## Required packages:

- bigstatsr
- bigsnpr
- lassosum
- data.table
- xgboost
- ranger
- keras

```{r, packages, warning=FALSE}
library(bigstatsr)
library(bigsnpr)
library(lassosum)
library(data.table)
library(xgboost)
library(ranger)
library(keras)
```

## Dataset

*Please change working direction in `directory`*

I use the "Simus" simulated dataset provided by Florian. This dataset contain ~650,000 SNPs in 2 chromosomes. 20% of them are cases and the rest 80% are controls.


```{r, load_data, warning=FALSE, message = FALSE}
setwd(directory) 
# Train data
sumstats <- bigreadr::fread2(file_sumstat)
# snp_readBed(paste(file_train, "bed", sep='.'))
train <- snp_attach(paste(file_train, "rds", sep='.'))
G.train <- train$genotypes
CHR <- train$map$chromosome
POS <- train$map$physical.pos
NCORES <- nb_cores()
lpval <- -log10(sumstats$pval)
y.train <- train$fam$affection

# Test data
# snp_readBed(paste(file_test, "bed", sep='.'))
test <- snp_attach(paste(file_test, "rds", sep='.'))
G.test <- test$genotypes
# 

```

## C+T step
Generating a matrix of C+T with varied clumping radius and $p-value$ thresholds.

$\beta s$ and $p-values$ are aquired in sumstats file (see `sumstats.txt`)

```{r, C+T,}
all_keep <- snp_grid_clumping(G.train, CHR, POS, lpval, ncores = NCORES)
PRS <-snp_grid_PRS(G.train,all_keep = all_keep, betas = sumstats$beta,lpval)
PRS_test <- snp_grid_PRS(G.test, all_keep = all_keep, betas = sumstats$beta,lpval)

```

Other than SCT, other methods require DataFrame or R Matrix than FBM to manage data. I converted the FBM output of C+T to DataFrame and change the corresponding columns.

```{r, include=FALSE}
# Convert to dataframe for non-FBM based methods
ds<-PRS[]
y.train <- train$fam$affection
PRS.df <- as.data.frame(ds)
cols <- colnames(PRS.df)
PRS.df$newcolumn <- y.train
colnames(PRS.df) <- append(cols, "y")
```

```{r include=FALSE,}
saveRDS(all_keep, "all_keep.rds")
all_keep <- readRDS("all_keep.rds")
```

## Data treatment: Oversampling

To oversampling I extract all positive (*affection = 1*), and append them a number of times. For this simulated dataset, I enhanced the number of cases 10 times, thus the ratio become ~70:30 (from 20:80).

```{r, Data-treatment, eval=FALSE}
y_case <- which(train$fam$affection %in% c(1))
for (i in seq(1, 10)){
  y.train <- append(y.train, train$fam$affection[y_case])
  ds <- rbind(ds, PRS[y_case,])
}
ds.FBM <- as_FBM(ds)
```

## Method 1: SCT

No parameters required. Combining (Sparse Logistic Regression) C+T predictors for best result.

```{r, SCT, results=FALSE}
# Reference
# M <- snp_grid_stacking(multi_PRS = ds.FBM, y.train = y.train, ncores = NCORES)
M <- snp_grid_stacking(multi_PRS = PRS, y.train = train$fam$affection)
beta <- as_FBM(matrix(M$beta.G))
pred.SCT <- big_prodMat(G.test, beta)
#AUC(pred = pred.SCT, test$fam$affection)
```
```{r,result_sct}
AUC(pred = pred.SCT, test$fam$affection)
```

## Method 2: XGBoost

XGBoost provide slightly better result compare to SCT. However this result heavily depend on booster and objective selection. Experiments shown that **gblinear** booster with **count:poisson** objective give the best results.

For experimentation, I run 9 different *max_depth* and 9 *nrounds*. Each combination is repeated 10 times and the mean value calculated to rule out the randomess of *gblinear* booster

```{r, xgboost, result=FALSE}
AUCs <- matrix(,nrow = 10, ncol = 10)
for (i in seq(2 : 10)){
  for (j in seq(2: 10)) {
    temp <- c()
    for(t in seq(1:10)){
      bstSparse <- xgboost(data = ds[], 
                           label = y.train, 
                           booster="gblinear",
                           max_depth = i, 
                           eta = 1, 
                           nthread = 2, 
                           nrounds = j, 
                           lambda = 0.1,
                           objective = "count:poisson"
                           )
      pred <- predict(bstSparse, PRS_test[])
      temp <- append(temp,AUC(pred = pred, test$fam$affection))
    }
    AUCs[i,j] <- mean(temp)
  }
}
```
```{r,result_xgboost}
AUCs
boxplot(AUCs)
```

## Method 3: Lassosum

Lassosum use L1 regularization to better fit the regression model.

For basic L1 regularization of Linear Regression, 2 metaparameters are required: a learning rate $\alpha$ and a regularization parameter $\lambda$. However, *lassosum* does not require these parameters, as the software automatically scan the parameters space and select the best $\alpha$ and $\lambda$

```{r, lassosum, result=FALSE}
setwd(directory) 
cor <- p2cor(p = sumstats$pval, n = 8000, sign=sumstats$beta)

out <- lassosum.pipeline(cor =cor, chr=sumstats$chromosome, pos = sumstats$physical.pos, 
                         A1 = sumstats$allele1, A2 = sumstats$allele2,
                         ref.bfile = file_train, test.bfile = file_test,
                         LDblocks = LDblocks)

v <- validate(out)

out2 <- subset(out, s=v$best.s, lambda = v$lambda)
v2 <- validate(out2)
v2$best.validation.result
```

```{r, result_lassosum}
AUC(v$best.pgs, v$pheno)
```

## Method 4: Random Forest

Random forest also give good AUC (above 70%, but less than other methods except for *keras*). The result highly dependen on the number of trees (more trees equals more convergence), and the learning rate also contribute to an lesser extend.

The code in this session will sweep through the parameters: 1 to 10 trees and alpha from 0.1 to 1 (0.1 increment).

```{r rf, result=FALSE}
PRS_test.df <- as.data.frame(PRS_test[])

f <- as.formula(paste("y ~", paste(cols[!cols %in% "PRS.df"], collapse = " + ")))

for (j in seq(1 : 10)){
  for (i in seq(1: 10)) {
    rf <- ranger(f, data = PRS.df, num.trees = j*100, verbose = TRUE, write.forest = TRUE, alpha = (i/10))
    pred <- predict(rf, PRS_test.df)
    AUCs[i,j] <- AUC(pred = pred$predictions, test$fam$affection)
  }
}
```

```{r, result_rf}
AUCs
boxplot(AUCs)
```

## Method 5: Nerual net (Keras)

For some reason, *keras* cannot return good result; the prediction always biased to controls. Even with adjusted weights and oversampling, the bias is still there and AUC is always $0.5$. I included it here in case I figure it out in the future.

In this example, I put a 10:1 weight ratio between case and control.

```{r, keras, result=FALSE}
y.train <- train$fam$affection

model <- keras_model_sequential() 
model %>%
 layer_dense(units = 1000, activation = "softmax", input_shape = c(ncol(ds)), kernel_regularizer=regularizer_l1(0.02)) %>%
 # layer_dense(units = 1000, activation = "softmax", input_shape = c(2800)) %>%
 # layer_dropout(rate=0.1) %>%
 layer_dense(units = 2, activation = "softmax", kernel_regularizer=regularizer_l1(0.01))

model %>% compile(
 #loss = 'categorical_crossentropy',
 loss = 'mean_absolute_percentage_error',
 optimizer = 'SGD',
 metrics = c('accuracy')
)
y.train <-  to_categorical(y.train)

history <- model$fit(
 ds, y.train,
 class_weight = list(1, 10),
 epochs = as.integer(10), 
 batch_size = as.integer(28)
)
```
```{r, result_keras}
AUC(predict_classes(model, PRS_test[]), test$fam$affection)

```
